### ISSUE TITLE
FileStream.Lock(Int64, Int64) always fails on FreeBSD

### ISSUE THREAD

**Issue by wfurt** [dotnet/runtime#65831]
wrapping discussion from  #65059, related to #14537.
Let's consider example as
```c#
using FileStream fs1 = File.Open(path, FileMode.Open, fileAccess, FileShare.ReadWrite);
fs1.Lock(0, 100);
fs1.Unlock(0, 100);
fs1.Lock(0, 100);
```

This will always fail with   something like
```System.IO.IOException: The process cannot access the file /tmp/FileStream_LockUnlock_nm3knbvl.5ix/OverlappingRegionsFromOtherProcess_With_WriteLock_ThrowsException_237_603d100b because it is being used by another process.``` 
The sequence above would translate to  following kernel calls:

```
16618: openat(AT_FDCWD,"/tmp/foo",O_RDWR|O_CLOEXEC,00) = 34 (0x22)
16618: flock(34,LOCK_SH|LOCK_NB)         = 0 (0x0)
16618: lseek(34,0x0,SEEK_CUR)            = 0 (0x0)
16618: lseek(34,0x0,SEEK_SET)            = 0 (0x0)
16618: fcntl(34,F_SETLK,0x7fffffffd928)      ERR#35 'Resource temporarily unavailable'
```
and 
```
     [EAGAIN]           The argument cmd is F_SETLK, the type of lock (l_type)
                        is a shared lock (F_RDLCK) or exclusive lock
                        (F_WRLCK), and the segment of a file to be locked is
                        already exclusive-locked by another process; or the
                        type is an exclusive lock and some portion of the
                        segment of a file to be locked is already shared-
                        locked or exclusive-locked by another process.
```

small `C` repro of the same calls:
```c
#include <sys/file.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>

int main(int argc, char** argv)
{
    struct flock fl;

    int fd= open("/tmp/foo", O_RDWR|O_CLOEXEC);
    int ret  = lseek(fd, 0, SEEK_SET);
    printf("Opened as %d\n", fd);
    if (argc > 1) {
        ret = flock(fd, LOCK_SH|LOCK_NB);
        printf("flock finished with %d\n", ret);
    }

    fl.l_start = 0;
    fl.l_len = 100;
    fl.l_type = F_WRLCK;
    fl.l_whence = SEEK_SET;

    ret = fcntl(fd, F_SETLK, &fl);
    printf("fcntl finished with %d and errno %d (%s)\n", ret, errno, strerror(errno));

    return 0;
}
```

It fails if we follow the `FileStream` sequence. 
But if we skip the `flock` then the `fcntl` works as expect. e.g. partial locking works but it seems like we cannot use both at the same time. 
On Linux, it works fine in both cases.

**Comment by adamsitnik** [comment]
@wfurt big thanks for all your help!

In such case I believe that we should simply throw `PlatformNotSupportedException` for `FileStream.Lock` and `FileStream.Unlock` similarly to what we do for OSX.

**Comment by adamsitnik** [comment]
## Up-for-grabs requirements

The contributor who is willing to work on this issue should:

- [ ] Extend the following two checks to throw on `FreeBSD` as well:

https://github.com/dotnet/runtime/blob/f97919dc4a527ca15e3eb4115046e02438bc9252/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.Unix.cs#L59-L64

https://github.com/dotnet/runtime/blob/f97919dc4a527ca15e3eb4115046e02438bc9252/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.Unix.cs#L69-L74

- [ ] annotate the `FileStream.Lock` and `Unlock` methods as unsupported on `FreeBSD`:

https://github.com/dotnet/runtime/blob/f97919dc4a527ca15e3eb4115046e02438bc9252/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs#L219-L222

https://github.com/dotnet/runtime/blob/f97919dc4a527ca15e3eb4115046e02438bc9252/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs#L236-L239

- [ ] do it for the reference assembly as well: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/ref/System.Runtime.cs

- [ ] find all `FileStream.Lock` and `Unlock` tests that are disabled for OSX and extend the condition with `FreeBSD` check, preferably by using `PlatformDetection.IsBsdLike` introduced by @wfurt in #65059

https://github.com/dotnet/runtime/blob/f97919dc4a527ca15e3eb4115046e02438bc9252/src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs#L47

- [ ] update the documentation

https://github.com/dotnet/dotnet-api-docs/blob/main/xml/System.IO/FileStream.xml

**Comment by emaste** [comment]
repro needs `O_CREAT` in the open() call, after adding that I get (on FreeBSD 14):
```
Opened as 3
fcntl finished with 0 and errno 0 (No error: 0)
```

**Comment by Tarun047** [comment]
Hi @adamsitnik , I'd like to take this one up, I've created a PR with the fixes mentioned above.
Kindly review and suggest if anything else is needed.

**Comment by adamsitnik** [comment]
Fixed by #65831

**Comment by adamsitnik** [comment]
@Tarun047 is there any chance you could update the docs as well? https://github.com/dotnet/dotnet-api-docs/blob/main/xml/System.IO/FileStream.xml
